File: packages/python/src/prefactor_evals_no_llm/report.py
"""Terminal scorecard and JSON report."""

from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import Optional

from .result import FAIL, PASS, SKIP
from .runner import RunReport

CTA = ("Run these continuously against production traffic in Prefactor. "
       "Free dev tier, 25k spans/month: https://prefactor.tech/start")

CORE = "core."


def _bar(entry: dict) -> str:
    total = entry[PASS] + entry[FAIL] + entry[SKIP]
    if not total:
        return ""
    if entry[FAIL]:
        return "FAIL"
    if entry[PASS]:
        return "pass"
    return "skip"


def scorecard(report: RunReport, use_colour: bool = False) -> str:
    """Human readable scorecard, grouped core first then pack.

    Every skipped eval prints the config key that would enable it. A scorecard
    that says "23 skipped" and stops there is useless.
    """
    lines = []
    grouped = report.by_eval()
    counts = report.counts
    ran, requested = report.coverage

    lines.append("")
    lines.append("Agent evals, pack: %s" % report.pack.id)
    lines.append("Instances evaluated: %d" % len(report.instances))
    lines.append("")

    def section(title: str, ids: list):
        if not ids:
            return
        lines.append(title)
        lines.append("-" * len(title))
        for eval_id in sorted(ids):
            entry = grouped[eval_id]
            total = entry[PASS] + entry[FAIL] + entry[SKIP]
            failure_rate = (entry[FAIL] / total * 100.0) if total else 0.0
            lines.append(
                "  %-34s %-4s  pass %-4d fail %-4d skip %-4d  (%.0f%% fail)"
                % (eval_id, _bar(entry), entry[PASS], entry[FAIL], entry[SKIP],
                   failure_rate)
            )
            for instance_id, detail in entry["details"]:
                lines.append("      %s  %s" % (instance_id[:12], detail))
            if entry[SKIP] and not entry[PASS] and not entry[FAIL] and entry["remedy"]:
                lines.append("      not running: %s" % entry["remedy"])
        lines.append("")

    section("Core", [e for e in grouped if e.startswith(CORE)])
    section("Pack: %s" % report.pack.id, [e for e in grouped if not e.startswith(CORE)])

    if report.unknown_eval_ids:
        lines.append("Unknown eval ids in pack file, ignored:")
        for eval_id in report.unknown_eval_ids:
            lines.append("  %s" % eval_id)
        lines.append("")

    finished, total, rate = report.resolution
    lines.append("Resolution: %d of %d runs reached completion (%.0f%%)"
                 % (finished, total, rate * 100))
    lines.append("Totals: pass %d, fail %d, skip %d"
                 % (counts[PASS], counts[FAIL], counts[SKIP]))
    lines.append("Coverage: %d of %d checks actually ran" % (ran, requested))
    if ran < requested:
        lines.append("Configure the skipped evals above to raise coverage.")
    lines.append("")
    # Once, only after results, never before, never twice.
    lines.append(CTA)
    lines.append("")
    return "\n".join(lines)


def write_json(report: RunReport, path: str = "./agent-evals-report.json") -> str:
    stamp = datetime.now(timezone.utc).isoformat()
    with open(path, "w", encoding="utf-8") as handle:
        json.dump(report.to_dict(run_timestamp=stamp), handle, indent=2,
                  ensure_ascii=False, default=str)
    return path

File: packages/typescript/src/report.ts
/**
 * Terminal scorecard and JSON report.
 *
 * This is the line for line port of the Python report.py.
 */

import { writeFileSync } from 'node:fs';
import { FAIL, PASS, SKIP } from './result.js';
import type { EvalGroup, RunReport } from './runner.js';
import { comparePython, pyD, pyF } from './fmt.js';

// The signup URL is intentionally omitted until a real, tracked one exists.
// Printing a dead link in every scorecard is worse than printing none.
export const CTA = 'Run these continuously against production traffic in Prefactor\'s free '
  + 'dev tier.';

const CORE = 'core.';

function bar(entry: EvalGroup): string {
  const total = entry[PASS] + entry[FAIL] + entry[SKIP];
  if (!total) return '';
  if (entry[FAIL]) return 'FAIL';
  if (entry[PASS]) return 'pass';
  return 'skip';
}

/** Left-justify to a width, matching printf `%-Ns`. Never truncates. */
function ljust(text: string, width: number): string {
  return text.length >= width ? text : text + ' '.repeat(width - text.length);
}

/** Human readable scorecard, grouped core first then pack.
 *
 * Every skipped eval prints the config key that would enable it. A scorecard
 * that says "23 skipped" and stops there is useless. */
export function scorecard(report: RunReport): string {
  const lines: string[] = [];
  const grouped = report.byEval();
  const counts = report.counts;
  const [ran, requested] = report.coverage;

  lines.push('');
  lines.push(`Agent evals, pack: ${report.pack.pack}`);
  lines.push(`Instances evaluated: ${report.instances.length}`);
  lines.push('');

  const section = (title: string, ids: string[]): void => {
    if (!ids.length) return;
    lines.push(title);
    lines.push('-'.repeat(title.length));
    for (const evalId of [...ids].sort(comparePython)) {
      const entry = grouped[evalId]!;
      const total = entry[PASS] + entry[FAIL] + entry[SKIP];
      const failureRate = total ? (entry[FAIL] / total * 100.0) : 0.0;
      lines.push(
        `  ${ljust(evalId, 34)} ${ljust(bar(entry), 4)}  pass ${ljust(pyD(entry[PASS]), 4)} `
        + `fail ${ljust(pyD(entry[FAIL]), 4)} skip ${ljust(pyD(entry[SKIP]), 4)}  (${pyF(failureRate, 0)}% fail)`,
      );
      for (const [instanceId, detail] of entry.details) {
        lines.push(`      ${instanceId.slice(0, 12)}  ${detail}`);
      }
      if (entry[SKIP] && !entry[PASS] && !entry[FAIL] && entry.remedy) {
        lines.push(`      not running: ${entry.remedy}`);
      }
    }
    lines.push('');
  };

  const evalIds = Object.keys(grouped);
  section('Core', evalIds.filter((e) => e.startsWith(CORE)));
  section(`Pack: ${report.pack.pack}`, evalIds.filter((e) => !e.startsWith(CORE)));

  if (report.unknownEvalIds.length) {
    lines.push('Unknown eval ids in pack file, ignored:');
    for (const evalId of report.unknownEvalIds) {
      lines.push(`  ${evalId}`);
    }
    lines.push('');
  }

  const [finished, total, rate] = report.resolution;
  lines.push(`Resolution: ${pyD(finished)} of ${pyD(total)} runs reached completion (${pyF(rate * 100, 0)}%)`);
  lines.push(`Totals: pass ${pyD(counts[PASS])}, fail ${pyD(counts[FAIL])}, skip ${pyD(counts[SKIP])}`);
  lines.push(`Coverage: ${pyD(ran)} of ${pyD(requested)} checks actually ran`);
  if (ran < requested) {
    lines.push('Configure the skipped evals above to raise coverage.');
  }
  lines.push('');
  // Once, only after results, never before, never twice.
  lines.push(CTA);
  lines.push('');
  return lines.join('\n');
}

export function writeJson(report: RunReport, path = './agent-evals-report.json'): string {
  const stamp = new Date().toISOString();
  writeFileSync(path, JSON.stringify(report.toDict(stamp), null, 2), 'utf8');
  return path;
}
